components/turret.py

```diff
@@ -1,42 +1,91 @@
+class Turret:
+    
+    goal_angle = tunable(0.0)
 
     def __init__(self) -> None:
         # Create the hardware object handles
-
+        self.motor = CANSparkMax(SparkMaxIds.turret_motor, CANSparkMax.MotorType.kBrushless) #This will need to change to the proper motor controller type
+        self.motor.setIdleMode(CANSparkMax.IdleMode.kBrake)
+        self.motor.setInverted(False)
+        self.positive_limit_switch = DigitalInput(DioChannels.positive_turret_switch)
+        self.negative_limit_switch= DigitalInput(DioChannels.negative_turret_switch) # These could also be attached directly to the motor controller breakout board for interrupts
+        self.encoder = self.motor.getEncoder()
+        self.encoder.setPositionConversionFactor(GEAR_RATIO)
+        self.encoder.setVelocityConversionFactor(GEAR_RATIO)
+        rotation_contraints = TrapezoidProfileRadians.Constraints(maxVelocity=MAX_ANGULAR_VELOCITY,maxAcceleration=MAX_ANGULAR_ACCELERATION)
+        self.rotation_controller = ProfiledPIDControllerRadians(2,0.0,0.0,rotation_contraints)
         # set index found var to false
-        pass
+        self.index_found = False
 
     def set_angle(self, angle: float) -> None:
         # set the desired angle for the turret
+        clamped_angle = min(angle, POSITIVE_LIMIT_ANGLE - radians(5))
+        clamped_angle = max(clamped_angle, NEGATIVE_LIMIT_ANGLE + radians(5))
+        self.goal_angle = clamped_angle        
 
-        # You will need to check the documentation if this can be on the output shaft
-        # or if you need to convert the angle back to a motor angle with the gear ratio
-        pass
+    @feedback
+    def get_angle(self) -> float:
+        return self.encoder.getPosition()
 
+    @feedback
     def at_angle(self) -> bool:
         # return abs(current angle - reference) < Tolerance
         # This could also be done inside the motor controller depending on how it works
-        return False
+        return abs(self.get_angle() - self.goal_angle) < ANGLE_ERROR_TOLERANCE
+
+    @feedback
+    def at_positive_limit(self) -> bool:
+        return not self.positive_limit_switch.get()
+    
+    @feedback
+    def at_negative_limit(self) -> bool:
+        return not self.negative_limit_switch.get()
 
     def find_index(self) -> None:
-        # configure turret to rotate until it finds an index
-        # probaly speed or voltage for this
-        pass
+        self.index_found = False
 
     def execute(self) -> None:
-        # if index found:
-        # calculate pid output based off angle delta
-        # calcualte feed forward based off angle delta
-        # clamp input
-        # set motor setpoint
-        # else:
+        if self.at_negative_limit():
+            self.encoder.setPosition(NEGATIVE_LIMIT_ANGLE)
+            self.index_found = True
+            
+
+        if self.at_positive_limit():
+            self.encoder.setPosition(POSITIVE_LIMIT_ANGLE)
+            self.index_found = True
+
+        
```

Comment:
We need code in here to avoid a repeat of "The Incident" - before these two checks we need to check if both limit switches are pressed simultaneously, and if so set the motor to not move and return from the function.